nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,222 @@
+              record_type: 'DatastoreRecord',
+              filter_func: Callable[[Union[Any, 'DatastoreRecord']], bool] = None,
+              filter_field: str = "",
+              writeable: bool = False,
+              ) -> List['DatastoreRecord']:
+        """
+        Performs a query on the datastore for the record by `record_type`.
+
+        An optional `filter_func` callable will take the decoded field
+        specified by the optional arg `filter_field` (see below) for the given
+        `record_type` iff the `filter_field` has been provided.
+        If no `filter_field` has been provided, then the `filter_func` will
+        receive a _readonly_ `DatastoreRecord`.
+
+        An optional `filter_field` can be provided as a `str` to perform a
+        query on a specific field for a `record_type`. This will cause the
+        `filter_func` to receive the decoded `filter_field` per the `record_type`.
+        Additionally, providing a `filter_field` will limit the query to
+        iterating over only the subset of records specific to that field.
+
+        If records can't be found, this method will raise `RecordNotFound`.
+        """
+        valid_records = set()
+        with self.__db_env.begin(write=writeable) as datastore_tx:
+            db_cursor = datastore_tx.cursor()
+
+            # Set the cursor to the closest key (if it exists) by the query params.
+            #
+            # By providing a `filter_field`, the query will immediately be
+            # limited to the subset of keys for the `filter_field`.
+            query_key = f'{record_type.__name__}:{filter_field}'.encode()
+            if not db_cursor.set_range(query_key):
+                # The cursor couldn't identify any records by the key
+                raise RecordNotFound(f"No records exist for the key from the specified query parameters: '{query_key}'")
+
+            # Check if the record at the cursor is valid for the query
+            curr_key = DatastoreKey.from_bytestring(db_cursor.key())
+            if not curr_key.compare_key(query_key):
+                raise RecordNotFound(f"No records exist for the key from the specified query parameters: '{query_key}'")
+
+            # Everything checks out, let's begin iterating!
+            # We begin by comparing the current key to the query key.
+            # If the key doesn't match the query key, we know that there are
+            # no records for the query because lmdb orders the keys lexicographically.
+            # Ergo, if the current key doesn't match the query key, we know
+            # we have gone beyond the relevant keys and can `break` the loop.
+            # Additionally, if the record is already in the `valid_records`
+            # set (identified by the `record_id`, we call `continue`.
+            for db_key in db_cursor.iternext(keys=True, values=False):
+                curr_key = DatastoreKey.from_bytestring(db_key)
+                if not curr_key.compare_key(query_key):
```

Comment:
After our discord discussion, I still think that  the name of this method should be related to `matches`, `conforms_to` or similar. I mean, even in the comment above you say "match".